include/xtensor-io/xio_binary.hpp

```diff
@@ -18,28 +18,69 @@ namespace xt
 {
     namespace detail
     {
+        bool is_big_endian(void)
+        {
+            union {
+                uint32_t i;
+                char c[4];
+            } bint = {0x01020304};
+
+            return bint.c[0] == 1;
+        }
+
+        template <class T>
+        void swap_endianness(const char* in_buffer, char* out_buffer, std::size_t size, T = T())
+        {
+            for (std::size_t i = 0; i < size; i++)
+            {
+                T tmp = *reinterpret_cast<const T*>(&(in_buffer[i * sizeof(T)]));
+                for (std::size_t j = 0; j < sizeof(T); j++)
+                {
+                    out_buffer[i * sizeof(T) + j] = reinterpret_cast<char*>(&tmp)[(i + 1) * sizeof(T) - j - 1];
+                }
+            }
+        }
+
         template <typename T>
-        inline xt::svector<T> load_bin_file(std::istream& stream)
+        inline xt::svector<T> load_bin_file(std::istream& stream, bool as_big_endian)
         {
             stream.seekg(0, stream.end);
             auto uncompressed_size = static_cast<std::size_t>(stream.tellg());
             stream.seekg(0, stream.beg);
-            xt::svector<T> uncompressed_buffer(uncompressed_size / sizeof(T));
-            stream.read(reinterpret_cast<char*>(uncompressed_buffer.data()), (std::streamsize)uncompressed_size);
+            std::size_t size = uncompressed_size / sizeof(T);
+            xt::svector<T> uncompressed_buffer(size);
+            char* buffer = reinterpret_cast<char*>(uncompressed_buffer.data());
+            stream.read(buffer, (std::streamsize)uncompressed_size);
+            if ((sizeof(T) > 1) && (as_big_endian != is_big_endian()))
+            {
+                swap_endianness<T>(buffer, buffer, size);
+            }
             return uncompressed_buffer;
         }
 
         template <class O, class E>
-        inline void dump_bin_stream(O& stream, const xexpression<E>& e)
+        inline void dump_bin_stream(O& stream, const xexpression<E>& e, bool as_big_endian)
         {
             using value_type = typename E::value_type;
             const E& ex = e.derived_cast();
             auto&& eval_ex = eval(ex);
             auto shape = eval_ex.shape();
-            std::size_t uncompressed_size = compute_size(shape) * sizeof(value_type);
+            std::size_t size = compute_size(shape);
+            std::size_t uncompressed_size = size * sizeof(value_type);
             const char* uncompressed_buffer = reinterpret_cast<const char*>(eval_ex.data());
-            stream.write(uncompressed_buffer, std::streamsize(uncompressed_size));
-            stream.flush();
+            if ((sizeof(value_type) > 1) && (as_big_endian != is_big_endian()))
+            {
+                char* swapped_buffer = new char[size * sizeof(value_type)];
```

Comment:
Please use `std::vector<char>` instead, it manages the memory for you and avoids memory leak if an exception is thrown.